pageserver/src/tenant.rs

```diff
@@ -948,11 +907,217 @@ impl Tenant {
+        // There is initially no data in it, but all the read-calls know to look
+        // into the ancestor.
+        let metadata = TimelineMetadata::new(
+            start_lsn,
+            dst_prev,
+            Some(src),
+            start_lsn,
+            *src_timeline.latest_gc_cutoff_lsn.read(),
+            src_timeline.initdb_lsn,
+        );
+        let new_timeline = self.create_initialized_timeline(dst, metadata, &mut timelines)?;
+        info!("branched timeline {dst} from {src} at {start_lsn}");
+
+        Ok(new_timeline)
+    }
+
+    /// - run initdb to init temporary instance and get bootstrap data
+    /// - after initialization complete, remove the temp dir.
+    fn bootstrap_timeline(&self, timeline_id: TimelineId) -> Result<Arc<Timeline>> {
+        // create a `tenant/{tenant_id}/timelines/basebackup-{timeline_id}.{TEMP_FILE_SUFFIX}/`
+        // temporary directory for basebackup files for the given timeline.
+        let initdb_path = path_with_suffix_extension(
+            self.conf
+                .timelines_path(&self.tenant_id)
+                .join(format!("basebackup-{timeline_id}")),
+            TEMP_FILE_SUFFIX,
+        );
+
+        // Init temporarily repo to get bootstrap data
+        run_initdb(self.conf, &initdb_path)?;
+        let pgdata_path = initdb_path;
+
+        let lsn = import_datadir::get_lsn_from_controlfile(&pgdata_path)?.align();
+
+        // Import the contents of the data directory at the initial checkpoint
+        // LSN, and any WAL after that.
+        // Initdb lsn will be equal to last_record_lsn which will be set after import.
+        // Because we know it upfront avoid having an option or dummy zero value by passing it to create_empty_timeline.
+        let timeline = self.create_empty_timeline(timeline_id, lsn)?;
+        import_datadir::import_timeline_from_postgres_datadir(&pgdata_path, &*timeline, lsn)?;
+
+        fail::fail_point!("before-checkpoint-new-timeline", |_| {
+            bail!("failpoint before-checkpoint-new-timeline");
+        });
+
+        timeline.checkpoint(CheckpointConfig::Forced)?;
+
+        info!(
+            "created root timeline {} timeline.lsn {}",
+            timeline_id,
+            timeline.get_last_record_lsn()
+        );
+
+        // Remove temp dir. We don't need it anymore
+        fs::remove_dir_all(pgdata_path)?;
+
+        Ok(timeline)
+    }
+
+    fn create_initialized_timeline(
+        &self,
+        new_timeline_id: TimelineId,
+        new_metadata: TimelineMetadata,
+        timelines: &mut MutexGuard<HashMap<TimelineId, Arc<Timeline>>>,
+    ) -> Result<Arc<Timeline>> {
+        crashsafe_dir::create_dir_all(self.conf.timeline_path(&new_timeline_id, &self.tenant_id))
+            .with_context(|| {
+            format!(
+                "Failed to create timeline {}/{} directory",
+                new_timeline_id, self.tenant_id
+            )
+        })?;
```

Comment:
One of the callers, `branch_timeline` does this already. Let's move this to `create_empty_timeline`, and add a comment that the caller is responsible for creating the timeline directory.

Reply:
Thank you for noticing.
I was actually trying to do the opposite and move in all FS-related methods into `create_initialized_timeline` so that we could later transfer it onto some "atomic" approach.

So, I'd leave it here and remove from the `branch_timeline` method, as I've done the same for `bootstrap_timeline`